### ISSUE TITLE
Make parsing XML from a file robustly easier, with a New-Xml cmdlet and the ability to cast file paths to [xml]

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#14505]
## Summary of the new feature/enhancement

* Implement a `New-Xml` cmdlet that robustly parses XML from a given file and returns an `[xml]` instance.

```powershell
# WISHFUL THINKING: 
# Should be the equivalent of:
#       & { $xmlDom = [xml]::new(); $xmlDoc.Load((Convert-Path file.xml)); return $xmlDom }
New-Xml file.xml   # -Path parameter implied.
```

 Like `Select-Xml`, a `-Content` parameter should support an XML _string_ as input.

* Complementarily, allow casts to `[xml]` to accept a single `Get-ChildItem` / `Get-Item` output object in order to parse from a file (which would complement the existing ability to cast a _string_; e.g., `[xml] '<xml><foo/></xml>'`)

```powershell
# WISHFUL THINKING: 
# Should be the equivalent of the above.
[xml] (Get-Item file.xml)
```

Note: We could even accept path _strings_, given that it's trivial to distinguish an XML _string_ (which has to start with `<`) from a file path.

---

**Rationale**:

Reading XML files into an `[xml]` DOM ([`System.Xml.XmlDocument`](https://docs.microsoft.com/en-US/dotnet/api/System.Xml.XmlDocument)) is a common use case, and the idiom frequently seen is:

```powershell
$xmlDom = [xml] (Get-Content -Raw file.xml)
```

While concise and convenient, especially compared to the robust alternative, this approach is _not robust_, because it doesn't respect character-encoding information that is _part of the file itself_, as part of the XML declaration - see [this Stack Overflow answer](https://stackoverflow.com/a/65264118/45375) for background.
As an aside: `Select-Xml` currently has the same problem - see #14404.

To currently get robust behavior, use of .NET APIs is required, which is far from obvious and cumbersome:

```powershell
# Robustly parse file 'file.xml' into an [xml] DOM
$xmlDom = [xml]::new(); $xmlDoc.Load((Convert-Path file.xml))
```

The proposed new features above would provide a  PowerShell-idiomatic alternative that is both robust and convenient.

---

Quick-and-dirty `New-Xml` prototype (for simplicity, a single parameter is used, and whether an XML string or a file path is given is derived from the specific value passed):

```powershell
function New-Xml {
  <#
  .SYNOPSIS
  Constructs an XML DOM ([xml] from an XML file or text.
  #>
  param([string] $PathOrText)
  Set-StrictMode -Version 1; $ErrorActionPreference = 'Stop'
  $doc = [xml]::new()
  if ($PathOrText[0] -eq '<') { $doc.LoadXml($PathOrText) }
  else                        { $doc.Load((Convert-Path $PathOrText)) } 
  return $doc 
}
```

**Comment by ThomasNieto** [comment]
I think breaking this functionality into two cmdlets would be better. `Import-Xml` to go along with the other `Import-*` cmdlets. Then update `ConvertTo-Xml` to support passing an XML formatted string (equivalent to [xml] '<xml><foo/></xml>').

**Comment by Tomalak** [comment]
All the XML cmdlets have the same problem as #14404, not only `Select-Xml`. Any XML document that is opened through the same internal function described over there suffers from this. Luckily that bug could be fixed without breaking existing interfaces (or scripts).

**Comment by mklement0** [comment]
@ThomasNieto, I can see the argument for following that pattern, but there's also the symmetry with `Select-Xml`, which I think should take precedence here.

Also, `ConvertTo-Xml` serves a different purpose (and, as an aside, is [virtually useless](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/6676) in its current form and I wish we deprecated it), which I think would be confusing.

As a philosophical aside, I've never liked how functionality is  divided across cmdlets along the lines of data from a file vs. in-memory data - for instance, why should a `ConvertFrom-Csv` cmdlet not also support loading from _files_, obviating the need for a separate `Import-Csv` cmdlet?

**Comment by Tomalak** [comment]
The sad reality is that the `[xml](Get-Content ...)` pattern is ubiquitous and there is no way fighting it, it just has been wrong for way too long.

Given that, one of the options would be to implement encoding sniffing for XML files in `Get-Content`, i.e. recreate the same mechanism that `XmlDocument` uses, at least as long as the user has not expressed a preference using then `-Encoding` parameter.

It would make things transparently correct for anyone copying code off of the Internet and who's not deep enough into the details of how XML implements file encodings. It's not a _clean_ solution, but a very pragmatic one. It also would help people who naively (or for performance reasons) process XML data line-wise as plain text.

**Comment by iRon7** [comment]
I like the idea of automatically distinguishing between a path and a `xml` string.
But there are a few (corner) cases to take care of:

### White spaces prefix
```PowerShell
New-Xml @'

<xml>
<foo/>
</xml>
'@
```

### Collection of string (lines)
```PowerShell
Get-Content .\data.xml | Select-Object Skip 2 | New-Xml
```

**Comment by mklement0** [comment]
That's a good point, @Tomalak, and I like the idea, but I think _both_ things are worth implementing.

That way, future users will have a more PowerShell-idiomatic way of constructing XML DOMs, whereas existing code and continued use of the old pattern implicitly benefits from the `Get-Content` enhancement.

Therefore, can I suggest you create a _new_ proposal based on your comment?
Some additional considerations:

* The precedence of inferring the encoding would have to be laid out: 
  * A BOM should take precedence over an XML declaration.
  * As currently, `-Encoding` should only be able to override interpretation of a file in the _absence_ of a BOM.
* Performance impact must be minimized.

**Comment by mklement0** [comment]
@iRon7:

If we implement this automatic distinction (which appeals to me), for string input I think it would be sufficient to do the following check:

* Is the first character a non-whitespace character other than `<`? If so, assume the input is a _file path_; otherwise, assume XML text, and pass the text as-is to `.LoadXml()`

Your `Get-Content` example is the type of command we want to get away from, at least given `Get-Content`'s current behavior of potentially misinterpreting the input file.

For consistency, I think it would be better to have `New-Xml` construct an `[xml]` instance for _each input object_ - be it a file path or text. Thus, piping text would require a single, multi-line string (expressed in terms of `Get-Content`: `Get-Content -Raw`):

**Comment by Tomalak** [comment]
An implementation for `Get-Content` encoding sniffing should 

- look at the first couple of bytes of the file as ASCII
- if there's a BOM use that (that's already happening)
- If it starts with the XML declaration `<?xml`, parse out the `encoding="..."`
  - if there is a recognized encoding name, switch the file stream to a new encoding
  - lacking that, default to UTF-8
- if the does not have an XML encoding, assume plain text in UTF-8
- not sure if viable: if there are decode errors, rewind stream, assume system default ANSI encoding

There's probably a few stumbling blocks along the way, but it's the general idea.

**Comment by mklement0** [comment]
Great, @Tomalak - can you please post this as a _new_ issue (of type "Feature Request/Idea")? We should continue the conversation there.

**Comment by Tomalak** [comment]
@mklement0 Done & moved details over.

**Comment by p0W3RH311** [comment]
@mklement0 why in your function prototype you use `Set-StrictMode -Version 1;` and not `Set-StrictMode -Version latest;`

**Comment by mklement0** [comment]
@p0W3RH311, because:

* you shouldn't use `Set-StrictMode -Version Latest` in production code, given that any PowerShell upgrade outside of your control can break that code.

* `Set-StrictMode -Version 2` and higher has a side effect that I find unacceptable (and which should have been fixed a long time ago): #2798
